File: backend/app/services/accounting_service.py
from typing import List, Optional, Dict, Any
from sqlalchemy.orm import Session
from datetime import date

from app.models.accounting import (
    ChartOfAccounts, JournalEntry, LedgerEntry, 
    FiscalPeriod, Payroll, PayrollItem, Employee
)
from app.schemas.accounting import (
    ChartOfAccountsCreate, ChartOfAccountsUpdate,
    JournalEntryCreate, JournalEntryUpdate,
    LedgerEntryCreate, LedgerEntryUpdate,
    FiscalPeriodCreate, FiscalPeriodUpdate,
    PayrollCreate, PayrollUpdate,
    PayrollItemCreate, PayrollItemUpdate,
    EmployeeCreate, EmployeeUpdate
)
from app.utils.id_generator import generate_journal_entry_number

class AccountingService:
    # Chart of Accounts methods
    @staticmethod
    def get_chart_of_accounts(db: Session, skip: int = 0, limit: int = 100) -> List[ChartOfAccounts]:
        """Get all chart of accounts with pagination"""
        return db.query(ChartOfAccounts).offset(skip).limit(limit).all()
    
    @staticmethod
    def get_account(db: Session, account_id: int) -> Optional[ChartOfAccounts]:
        """Get a single account by ID"""
        return db.query(ChartOfAccounts).filter(ChartOfAccounts.id == account_id).first()
    
    @staticmethod
    def get_account_by_number(db: Session, account_number: str) -> Optional[ChartOfAccounts]:
        """Get a single account by account number"""
        return db.query(ChartOfAccounts).filter(ChartOfAccounts.account_number == account_number).first()
    
    @staticmethod
    def create_account(db: Session, account: ChartOfAccountsCreate) -> ChartOfAccounts:
        """Create a new account"""
        db_account = ChartOfAccounts(**account.dict())
        db.add(db_account)
        db.commit()
        db.refresh(db_account)
        return db_account
    
    @staticmethod
    def update_account(db: Session, account_id: int, account: ChartOfAccountsUpdate) -> Optional[ChartOfAccounts]:
        """Update an existing account"""
        db_account = AccountingService.get_account(db, account_id)
        if db_account:
            update_data = account.dict(exclude_unset=True)
            for key, value in update_data.items():
                setattr(db_account, key, value)
            
            db.commit()
            db.refresh(db_account)
        return db_account
    
    @staticmethod
    def delete_account(db: Session, account_id: int) -> bool:
        """Delete an account"""
        db_account = AccountingService.get_account(db, account_id)
        if db_account:
            db.delete(db_account)
            db.commit()
            return True
        return False
    
    # Journal Entry methods
    @staticmethod
    def get_journal_entries(db: Session, skip: int = 0, limit: int = 100) -> List[JournalEntry]:
        """Get all journal entries with pagination"""
        return db.query(JournalEntry).offset(skip).limit(limit).all()
    
    @staticmethod
    def get_journal_entry(db: Session, entry_id: int) -> Optional[JournalEntry]:
        """Get a single journal entry by ID"""
        return db.query(JournalEntry).filter(JournalEntry.id == entry_id).first()
    
    @staticmethod
    def get_journal_entry_by_number(db: Session, entry_number: str) -> Optional[JournalEntry]:
        """Get a single journal entry by entry number"""
        return db.query(JournalEntry).filter(JournalEntry.entry_number == entry_number).first()
    
    @staticmethod
    def create_journal_entry(db: Session, entry: JournalEntryCreate) -> JournalEntry:
        """Create a new journal entry"""
        # If entry_number is not provided, generate one
        if not entry.entry_number:
            entry_dict = entry.dict()
            entry_dict["entry_number"] = generate_journal_entry_number(db)
            db_entry = JournalEntry(**entry_dict)
        else:
            db_entry = JournalEntry(**entry.dict())
        
        db.add(db_entry)
        db.commit()
        db.refresh(db_entry)
        return db_entry
    
    @staticmethod
    def update_journal_entry(db: Session, entry_id: int, entry: JournalEntryUpdate) -> Optional[JournalEntry]:
        """Update an existing journal entry"""
        db_entry = AccountingService.get_journal_entry(db, entry_id)
        if db_entry:
            update_data = entry.dict(exclude_unset=True)
            for key, value in update_data.items():
                setattr(db_entry, key, value)
            
            db.commit()
            db.refresh(db_entry)
        return db_entry
    
    @staticmethod
    def delete_journal_entry(db: Session, entry_id: int) -> bool:
        """Delete a journal entry"""
        db_entry = AccountingService.get_journal_entry(db, entry_id)
        if db_entry:
            db.delete(db_entry)
            db.commit()
            return True
        return False
    
    # Ledger Entry methods
    @staticmethod
    def get_ledger_entries(db: Session, journal_entry_id: int) -> List[LedgerEntry]:
        """Get all ledger entries for a journal entry"""
        return db.query(LedgerEntry).filter(LedgerEntry.journal_entry_id == journal_entry_id).all()
    
    @staticmethod
    def get_ledger_entry(db: Session, entry_id: int) -> Optional[LedgerEntry]:
        """Get a single ledger entry by ID"""
        return db.query(LedgerEntry).filter(LedgerEntry.id == entry_id).first()
    
    @staticmethod
    def create_ledger_entry(db: Session, entry: LedgerEntryCreate) -> LedgerEntry:
        """Create a new ledger entry"""
        db_entry = LedgerEntry(**entry.dict())
        db.add(db_entry)
        db.commit()
        db.refresh(db_entry)
        return db_entry
    
    @staticmethod
    def update_ledger_entry(db: Session, entry_id: int, entry: LedgerEntryUpdate) -> Optional[LedgerEntry]:
        """Update an existing ledger entry"""
        db_entry = AccountingService.get_ledger_entry(db, entry_id)
        if db_entry:
            update_data = entry.dict(exclude_unset=True)
            for key, value in update_data.items():
                setattr(db_entry, key, value)
            
            db.commit()
            db.refresh(db_entry)
        return db_entry
    
    @staticmethod
    def delete_ledger_entry(db: Session, entry_id: int) -> bool:
        """Delete a ledger entry"""
        db_entry = AccountingService.get_ledger_entry(db, entry_id)
        if db_entry:
            db.delete(db_entry)
            db.commit()
            return True
        return False
    
    # Fiscal Period methods
    @staticmethod
    def get_fiscal_periods(db: Session, skip: int = 0, limit: int = 100) -> List[FiscalPeriod]:
        """Get all fiscal periods with pagination"""
        return db.query(FiscalPeriod).offset(skip).limit(limit).all()
    
    @staticmethod
    def get_fiscal_period(db: Session, period_id: int) -> Optional[FiscalPeriod]:
        """Get a single fiscal period by ID"""
        return db.query(FiscalPeriod).filter(FiscalPeriod.id == period_id).first()
    
    @staticmethod
    def create_fiscal_period(db: Session, period: FiscalPeriodCreate) -> FiscalPeriod:
        """Create a new fiscal period"""
        db_period = FiscalPeriod(**period.dict())
        db.add(db_period)
        db.commit()
        db.refresh(db_period)
        return db_period
    
    @staticmethod
    def update_fiscal_period(db: Session, period_id: int, period: FiscalPeriodUpdate) -> Optional[FiscalPeriod]:
        """Update an existing fiscal period"""
        db_period = AccountingService.get_fiscal_period(db, period_id)
        if db_period:
            update_data = period.dict(exclude_unset=True)
            for key, value in update_data.items():
                setattr(db_period, key, value)
            
            db.commit()
            db.refresh(db_period)
        return db_period
    
    @staticmethod
    def delete_fiscal_period(db: Session, period_id: int) -> bool:
        """Delete a fiscal period"""
        db_period = AccountingService.get_fiscal_period(db, period_id)
        if db_period:
            db.delete(db_period)
            db.commit()
            return True
        return False
    
    # Payroll methods
    @staticmethod
    def get_payrolls(db: Session, skip: int = 0, limit: int = 100) -> List[Payroll]:
        """Get all payrolls with pagination"""
        return db.query(Payroll).offset(skip).limit(limit).all()
    
    @staticmethod
    def get_payroll(db: Session, payroll_id: int) -> Optional[Payroll]:
        """Get a single payroll by ID"""
        return db.query(Payroll).filter(Payroll.id == payroll_id).first()
    
    @staticmethod
    def create_payroll(db: Session, payroll: PayrollCreate) -> Payroll:
        """Create a new payroll"""
        db_payroll = Payroll(**payroll.dict())
        db.add(db_payroll)
        db.commit()
        db.refresh(db_payroll)
        return db_payroll
    
    @staticmethod
    def update_payroll(db: Session, payroll_id: int, payroll: PayrollUpdate) -> Optional[Payroll]:
        """Update an existing payroll"""
        db_payroll = AccountingService.get_payroll(db, payroll_id)
        if db_payroll:
            update_data = payroll.dict(exclude_unset=True)
            for key, value in update_data.items():
                setattr(db_payroll, key, value)
            
            db.commit()
            db.refresh(db_payroll)
        return db_payroll
    
    @staticmethod
    def delete_payroll(db: Session, payroll_id: int) -> bool:
        """Delete a payroll"""
        db_payroll = AccountingService.get_payroll(db, payroll_id)
        if db_payroll:
            db.delete(db_payroll)
            db.commit()
            return True
        return False
    
    # Payroll Item methods
    @staticmethod
    def get_payroll_items(db: Session, payroll_id: int) -> List[PayrollItem]:
        """Get all payroll items for a payroll"""
        return db.query(PayrollItem).filter(PayrollItem.payroll_id == payroll_id).all()
    
    @staticmethod
    def get_payroll_item(db: Session, item_id: int) -> Optional[PayrollItem]:
        """Get a single payroll item by ID"""
        return db.query(PayrollItem).filter(PayrollItem.id == item_id).first()
    
    @staticmethod
    def create_payroll_item(db: Session, item: PayrollItemCreate) -> PayrollItem:
        """Create a new payroll item"""
        db_item = PayrollItem(**item.dict())
        db.add(db_item)
        db.commit()
        db.refresh(db_item)
        
        # Update the payroll's total amount
        payroll = AccountingService.get_payroll(db, item.payroll_id)
        if payroll:
            payroll.total_amount += item.net_salary
            db.commit()
        
        return db_item
    
    @staticmethod
    def update_payroll_item(db: Session, item_id: int, item: PayrollItemUpdate) -> Optional[PayrollItem]:
        """Update an existing payroll item"""
        db_item = AccountingService.get_payroll_item(db, item_id)
        if db_item:
            # Store the old net salary for updating the payroll total
            old_net_salary = db_item.net_salary
            
            update_data = item.dict(exclude_unset=True)
            for key, value in update_data.items():
                setattr(db_item, key, value)
            
            db.commit()
            db.refresh(db_item)
            
            # Update the payroll's total amount if net_salary was updated
            if "net_salary" in update_data:
                payroll = AccountingService.get_payroll(db, db_item.payroll_id)
                if payroll:
                    payroll.total_amount = payroll.total_amount - old_net_salary + db_item.net_salary
                    db.commit()
        
        return db_item
    
    @staticmethod
    def delete_payroll_item(db: Session, item_id: int) -> bool:
        """Delete a payroll item"""
        db_item = AccountingService.get_payroll_item(db, item_id)
        if db_item:
            # Update the payroll's total amount
            payroll = AccountingService.get_payroll(db, db_item.payroll_id)
            if payroll:
                payroll.total_amount -= db_item.net_salary
                db.commit()
            
            db.delete(db_item)
            db.commit()
            return True
        return False
    
    # Employee methods
    @staticmethod
    def get_employees(db: Session, skip: int = 0, limit: int = 100) -> List[Employee]:
        """Get all employees with pagination"""
        return db.query(Employee).offset(skip).limit(limit).all()
    
    @staticmethod
    def get_employee(db: Session, employee_id: int) -> Optional[Employee]:
        """Get a single employee by ID"""
        return db.query(Employee).filter(Employee.id == employee_id).first()
    
    @staticmethod
    def get_employee_by_employee_id(db: Session, employee_id: str) -> Optional[Employee]:
        """Get a single employee by employee ID"""
        return db.query(Employee).filter(Employee.employee_id == employee_id).first()
    
    @staticmethod
    def create_employee(db: Session, employee: EmployeeCreate) -> Employee:
        """Create a new employee"""
        db_employee = Employee(**employee.dict())
        db.add(db_employee)
        db.commit()
        db.refresh(db_employee)
        return db_employee
    
    @staticmethod
    def update_employee(db: Session, employee_id: int, employee: EmployeeUpdate) -> Optional[Employee]:
        """Update an existing employee"""
        db_employee = AccountingService.get_employee(db, employee_id)
        if db_employee:
            update_data = employee.dict(exclude_unset=True)
            for key, value in update_data.items():
                setattr(db_employee, key, value)
            
            db.commit()
            db.refresh(db_employee)
        return db_employee
    
    @staticmethod
    def delete_employee(db: Session, employee_id: int) -> bool:
        """Delete an employee"""
        db_employee = AccountingService.get_employee(db, employee_id)
        if db_employee:
            db.delete(db_employee)
            db.commit()
            return True
        return False

File: frontend/src/services/accountingService.js
import api from './api';

/**
 * Accounting Service
 * 
 * This service handles all API calls related to accounting.
 */
const accountingService = {
  /**
   * Chart of Accounts methods
   */
  getChartOfAccounts: async (skip = 0, limit = 100) => {
    try {
      const response = await api.get(`/accounting/chart-of-accounts?skip=${skip}&limit=${limit}`);
      return response.data;
    } catch (error) {
      console.error('Error fetching chart of accounts:', error);
      throw error;
    }
  },

  getAccount: async (id) => {
    try {
      const response = await api.get(`/accounting/chart-of-accounts/${id}`);
      return response.data;
    } catch (error) {
      console.error(`Error fetching account with ID ${id}:`, error);
      throw error;
    }
  },

  createAccount: async (accountData) => {
    try {
      const response = await api.post('/accounting/chart-of-accounts', accountData);
      return response.data;
    } catch (error) {
      console.error('Error creating account:', error);
      throw error;
    }
  },

  updateAccount: async (id, accountData) => {
    try {
      const response = await api.put(`/accounting/chart-of-accounts/${id}`, accountData);
      return response.data;
    } catch (error) {
      console.error(`Error updating account with ID ${id}:`, error);
      throw error;
    }
  },

  deleteAccount: async (id) => {
    try {
      const response = await api.delete(`/accounting/chart-of-accounts/${id}`);
      return response.data;
    } catch (error) {
      console.error(`Error deleting account with ID ${id}:`, error);
      throw error;
    }
  },

  /**
   * Journal Entry methods
   */
  getJournalEntries: async (skip = 0, limit = 100) => {
    try {
      const response = await api.get(`/accounting/journal-entries?skip=${skip}&limit=${limit}`);
      return response.data;
    } catch (error) {
      console.error('Error fetching journal entries:', error);
      throw error;
    }
  },

  getJournalEntry: async (id) => {
    try {
      const response = await api.get(`/accounting/journal-entries/${id}`);
      return response.data;
    } catch (error) {
      console.error(`Error fetching journal entry with ID ${id}:`, error);
      throw error;
    }
  },

  createJournalEntry: async (entryData) => {
    try {
      const response = await api.post('/accounting/journal-entries', entryData);
      return response.data;
    } catch (error) {
      console.error('Error creating journal entry:', error);
      throw error;
    }
  },

  updateJournalEntry: async (id, entryData) => {
    try {
      const response = await api.put(`/accounting/journal-entries/${id}`, entryData);
      return response.data;
    } catch (error) {
      console.error(`Error updating journal entry with ID ${id}:`, error);
      throw error;
    }
  },

  deleteJournalEntry: async (id) => {
    try {
      const response = await api.delete(`/accounting/journal-entries/${id}`);
      return response.data;
    } catch (error) {
      console.error(`Error deleting journal entry with ID ${id}:`, error);
      throw error;
    }
  },

  /**
   * Ledger Entry methods
   */
  getLedgerEntries: async (journalEntryId) => {
    try {
      const response = await api.get(`/accounting/journal-entries/${journalEntryId}/ledger-entries`);
      return response.data;
    } catch (error) {
      console.error(`Error fetching ledger entries for journal entry with ID ${journalEntryId}:`, error);
      throw error;
    }
  },

  getLedgerEntry: async (id) => {
    try {
      const response = await api.get(`/accounting/ledger-entries/${id}`);
      return response.data;
    } catch (error) {
      console.error(`Error fetching ledger entry with ID ${id}:`, error);
      throw error;
    }
  },

  createLedgerEntry: async (journalEntryId, entryData) => {
    try {
      const response = await api.post(`/accounting/journal-entries/${journalEntryId}/ledger-entries`, {
        ...entryData,
        journal_entry_id: journalEntryId
      });
      return response.data;
    } catch (error) {
      console.error(`Error creating ledger entry for journal entry with ID ${journalEntryId}:`, error);
      throw error;
    }
  },

  updateLedgerEntry: async (id, entryData) => {
    try {
      const response = await api.put(`/accounting/ledger-entries/${id}`, entryData);
      return response.data;
    } catch (error) {
      console.error(`Error updating ledger entry with ID ${id}:`, error);
      throw error;
    }
  },

  deleteLedgerEntry: async (id) => {
    try {
      const response = await api.delete(`/accounting/ledger-entries/${id}`);
      return response.data;
    } catch (error) {
      console.error(`Error deleting ledger entry with ID ${id}:`, error);
      throw error;
    }
  },

  /**
   * Fiscal Period methods
   */
  getFiscalPeriods: async (skip = 0, limit = 100) => {
    try {
      const response = await api.get(`/accounting/fiscal-periods?skip=${skip}&limit=${limit}`);
      return response.data;
    } catch (error) {
      console.error('Error fetching fiscal periods:', error);
      throw error;
    }
  },

  getFiscalPeriod: async (id) => {
    try {
      const response = await api.get(`/accounting/fiscal-periods/${id}`);
      return response.data;
    } catch (error) {
      console.error(`Error fetching fiscal period with ID ${id}:`, error);
      throw error;
    }
  },

  createFiscalPeriod: async (periodData) => {
    try {
      const response = await api.post('/accounting/fiscal-periods', periodData);
      return response.data;
    } catch (error) {
      console.error('Error creating fiscal period:', error);
      throw error;
    }
  },

  updateFiscalPeriod: async (id, periodData) => {
    try {
      const response = await api.put(`/accounting/fiscal-periods/${id}`, periodData);
      return response.data;
    } catch (error) {
      console.error(`Error updating fiscal period with ID ${id}:`, error);
      throw error;
    }
  },

  deleteFiscalPeriod: async (id) => {
    try {
      const response = await api.delete(`/accounting/fiscal-periods/${id}`);
      return response.data;
    } catch (error) {
      console.error(`Error deleting fiscal period with ID ${id}:`, error);
      throw error;
    }
  },

  /**
   * Payroll methods
   */
  getPayrolls: async (skip = 0, limit = 100) => {
    try {
      const response = await api.get(`/accounting/payrolls?skip=${skip}&limit=${limit}`);
      return response.data;
    } catch (error) {
      console.error('Error fetching payrolls:', error);
      throw error;
    }
  },

  getPayroll: async (id) => {
    try {
      const response = await api.get(`/accounting/payrolls/${id}`);
      return response.data;
    } catch (error) {
      console.error(`Error fetching payroll with ID ${id}:`, error);
      throw error;
    }
  },

  createPayroll: async (payrollData) => {
    try {
      const response = await api.post('/accounting/payrolls', payrollData);
      return response.data;
    } catch (error) {
      console.error('Error creating payroll:', error);
      throw error;
    }
  },

  updatePayroll: async (id, payrollData) => {
    try {
      const response = await api.put(`/accounting/payrolls/${id}`, payrollData);
      return response.data;
    } catch (error) {
      console.error(`Error updating payroll with ID ${id}:`, error);
      throw error;
    }
  },

  deletePayroll: async (id) => {
    try {
      const response = await api.delete(`/accounting/payrolls/${id}`);
      return response.data;
    } catch (error) {
      console.error(`Error deleting payroll with ID ${id}:`, error);
      throw error;
    }
  },

  /**
   * Payroll Item methods
   */
  getPayrollItems: async (payrollId) => {
    try {
      const response = await api.get(`/accounting/payrolls/${payrollId}/items`);
      return response.data;
    } catch (error) {
      console.error(`Error fetching payroll items for payroll with ID ${payrollId}:`, error);
      throw error;
    }
  },

  getPayrollItem: async (id) => {
    try {
      const response = await api.get(`/accounting/payroll-items/${id}`);
      return response.data;
    } catch (error) {
      console.error(`Error fetching payroll item with ID ${id}:`, error);
      throw error;
    }
  },

  createPayrollItem: async (payrollId, itemData) => {
    try {
      const response = await api.post(`/accounting/payrolls/${payrollId}/items`, {
        ...itemData,
        payroll_id: payrollId
      });
      return response.data;
    } catch (error) {
      console.error(`Error creating payroll item for payroll with ID ${payrollId}:`, error);
      throw error;
    }
  },

  updatePayrollItem: async (id, itemData) => {
    try {
      const response = await api.put(`/accounting/payroll-items/${id}`, itemData);
      return response.data;
    } catch (error) {
      console.error(`Error updating payroll item with ID ${id}:`, error);
      throw error;
    }
  },

  deletePayrollItem: async (id) => {
    try {
      const response = await api.delete(`/accounting/payroll-items/${id}`);
      return response.data;
    } catch (error) {
      console.error(`Error deleting payroll item with ID ${id}:`, error);
      throw error;
    }
  },

  /**
   * Employee methods
   */
  getEmployees: async (skip = 0, limit = 100) => {
    try {
      const response = await api.get(`/accounting/employees?skip=${skip}&limit=${limit}`);
      return response.data;
    } catch (error) {
      console.error('Error fetching employees:', error);
      throw error;
    }
  },

  getEmployee: async (id) => {
    try {
      const response = await api.get(`/accounting/employees/${id}`);
      return response.data;
    } catch (error) {
      console.error(`Error fetching employee with ID ${id}:`, error);
      throw error;
    }
  },

  createEmployee: async (employeeData) => {
    try {
      const response = await api.post('/accounting/employees', employeeData);
      return response.data;
    } catch (error) {
      console.error('Error creating employee:', error);
      throw error;
    }
  },

  updateEmployee: async (id, employeeData) => {
    try {
      const response = await api.put(`/accounting/employees/${id}`, employeeData);
      return response.data;
    } catch (error) {
      console.error(`Error updating employee with ID ${id}:`, error);
      throw error;
    }
  },

  deleteEmployee: async (id) => {
    try {
      const response = await api.delete(`/accounting/employees/${id}`);
      return response.data;
    } catch (error) {
      console.error(`Error deleting employee with ID ${id}:`, error);
      throw error;
    }
  }
};

export default accountingService;
